openstack-mixin/panels.libsonnet

```diff
@@ -0,0 +1,1004 @@
+        'Projects',
+        targets=[t.projects],
+        description='The number of projects for the OpenStack cloud.',
+      )
+      + stat.options.withGraphMode('none'),
+
+    regions:
+      commonlib.panels.generic.stat.info.new(
+        'Regions',
+        targets=[t.regions],
+        description='The number of regions for the OpenStack cloud.',
+      )
+      + stat.options.withGraphMode('none'),
+
+    users:
+      commonlib.panels.generic.timeSeries.base.new(
+        'Users',
+        targets=[t.users],
+        description='The number of users for the OpenStack cloud.',
+      )
+      + timeSeries.standardOptions.withDecimals(0),
+
+    projectDetails:
+      table.new('Project details')
+      + table.queryOptions.withTargets([t.projectDetails])
+      + table.queryOptions.withDatasource('prometheus', '${prometheus_datasource}')
+      + table.panelOptions.withDescription('Details for the projects in the OpenStack cloud.')
+      + table.standardOptions.withOverridesMixin([
+        table.fieldOverride.byName.new('Enabled')
+        + table.fieldOverride.byName.withProperty('custom.displayMode', 'color-text')
+        + table.fieldOverride.byName.withPropertiesFromOptions(
+          table.standardOptions.withMappings(
+            {
+              type: 'value',
+              options: {
+                'false': {
+                  color: 'red',
+                  index: 1,
+                  text: 'False',
+                },
+                'true': {
+                  color: 'green',
+                  index: 0,
+                  text: 'True',
+                },
+              },
+            }
+          ),
+        ),
+      ])
+      + table.queryOptions.withTransformationsMixin([
+        {
+          id: 'joinByField',
+          options: {
+            byField: 'Time',
+            mode: 'outer',
+          },
+        },
+        {
+          id: 'organize',
+          options: {
+            excludeByName: {
+              Time: true,
+              Value: true,
+              __name__: true,
+              description: true,
+              job: true,
+              parent_id: true,
+            },
+            includeByName: {},
+            indexByName: {
+              Time: 3,
+              Value: 11,
+              __name__: 4,
+              description: 5,
+              domain_id: 7,
+              enabled: 6,
+              id: 1,
+              instance: 2,
+              is_domain: 8,
+              job: 10,
+              name: 0,
+              parent_id: 9,
+            },
+            renameByName: {
+              domain_id: 'Domain ID',
+              enabled: 'Enabled',
+              id: 'ID',
+              instance: 'Instance',
+              is_domain: 'Is domain',
+              name: 'Name',
+            },
+          },
+        },
+      ]),
+
+    vms:
+      commonlib.panels.generic.timeSeries.base.new(
+        'VMs',
+        targets=[t.vms],
```

Comment:
Looks like compared to the prototype we went down to a single query vs. 2 with legend calling out " - total" and " - running"?

Reply:
Yeah. The metric for running vms does exist, but the exporter fails to report it even after applying load gen. There is an error output every time the exporter is called for that metric, so I excluded it going forward and just used the 'total vms' metric.